src/catalog/src/remote/cached_client.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+use std::sync::Arc;
+use std::time::Duration;
+
+use common_telemetry::timer;
+use meta_client::client::MetaClient;
+use moka::future::{Cache, CacheBuilder};
+
+use super::MetaKvBackend;
+use crate::error::Error::NoSuchKey;
+use crate::error::{Error, GenericSnafu, NoSuchKeySnafu, Result};
+use crate::metrics::{METRIC_CATALOG_KV_GET, METRIC_CATALOG_KV_REMOTE_GET};
+use crate::remote::{CachedKvBackend, Kv, KvBackend, KvBackendRef, ValueIter};
+
+pub struct CachedCatalogKvBackend {
+    inner_kv_backend: KvBackendRef,
+    cache: Arc<Cache<Vec<u8>, Kv>>,
+}
+
+#[async_trait::async_trait]
+impl KvBackend for CachedCatalogKvBackend {
+    fn as_any(&self) -> &(dyn Any + Send + Sync) {
+        self
+    }
+
+    fn range<'a, 'b>(&'a self, key: &[u8]) -> ValueIter<'b, Error>
+    where
+        'a: 'b,
+    {
+        self.inner_kv_backend.range(key)
+    }
+
+    async fn get(&self, key: &[u8]) -> Result<Option<Kv>> {
+        let _timer = timer!(METRIC_CATALOG_KV_GET);
+
+        let init = async {
+            let _timer = timer!(METRIC_CATALOG_KV_REMOTE_GET);
+
+            let kv = self.inner_kv_backend.get(key).await;
+            convert(kv)
```

Comment:
I think it's better to cache the `None` (wrap it with some concrete struct) value as well, to avoid cache penetration.

Reply:
Agree. But we have to take care of the cache validation. If the result is present, we should invalidate the `None` cache item too.